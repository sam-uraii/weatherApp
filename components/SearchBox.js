import { SearchBar } from "@rneui/themed";
import React from "react";

const SearchBox = ({
  updateSearch,
  searchedKeyword,
  onClearCallBack,
  setInputRef,
  isSearchLoading,
  onFocus,
}) => {
  return (
    <SearchBar
      ref={setInputRef}
      inputContainerStyle={{ height: 40 }}
      onFocus={() => onFocus()}
      placeholder="City Name..."
      onChangeText={updateSearch}
      value={searchedKeyword}
      lightTheme={true}
      onClear={onClearCallBack}
      round={true}
      showLoading={isSearchLoading}
    />
  );
};
export default SearchBox;
