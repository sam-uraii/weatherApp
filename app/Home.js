import { ActivityIndicator, View, Appearance, StyleSheet } from "react-native";
import { useEffect, useRef, useState } from "react";
import SearchedCityList from "../Components/SearchedCityList";
import { connect } from "react-redux";
import { fetchSearchedCities } from "../Redux/Middleware/searchedCityMiddleware";
import {
  updateSearchedKeyword,
  clearSearchedCityReducer,
} from "../Redux/Action/searchedCityAction";
import CurrentWeatherInfo from "../Components/Weather/CurrentWeatherInfo";
import WeatherForecastDaysList from "../Components/Weather/WeatherForecastDaysList";
import DropDown from "../Components/DropDown";
import { FORECAST_DAYS } from "../Constants/DropdownConstants";
import { updateNumberOfDaysForForecast } from "../Redux/Action/weatherDetailsOfCityAction";
import { getWeatherForecastDetailsOfCity } from "../Redux/Middleware/weatherDetailsOfCityMiddleware";
import { defaultBackgroundColor } from "../Constants/Colors";
import SearchBox from "../Components/SearchBox";
const Home = ({
  fetchSearchedCities,
  updateSearchedKeyword,
  searchedKeyword,
  searchedCities,
  clearSearchedCityReducer,
  isSearchLoading,
  selectedCity,
  weatherDetails,
  isWeatherDetailsLoading,
  weatherForecastDetails,
  isWeatherForecastDetailsLoading,
  updateNumberOfDaysForForecast,
  numberOfDaysForForecast,
  getWeatherForecastDetailsOfCity,
}) => {
  const [areWeatherDetailsVisible, setAreWeatherDetailsVisible] =
    useState(false);
  const [isSearchedCityListVisible, setIsSearchedCityListVisible] =
    useState(false);
  let inputRef = useRef(null);
  useEffect(() => {
    if (selectedCity) {
      getWeatherForecastDetailsOfCity();
    }
  }, [numberOfDaysForForecast]);
  useEffect(() => {
    if (selectedCity) {
      inputRef.blur();
      setAreWeatherDetailsVisible(true);
      setIsSearchedCityListVisible(false);
    }
  }, [selectedCity]);
  useEffect(() => {
    if (searchedCities.length) {
      setIsSearchedCityListVisible(true);
    } else {
      setIsSearchedCityListVisible(false);
    }
  }, [searchedCities]);

  const updateSearch = async (keyword) => {
    updateSearchedKeyword(keyword);
    fetchSearchedCities();
  };

  const onFocus = () => {
    if (searchedCities.length) {
      setIsSearchedCityListVisible(true);
    }
  };
  const onClearCallBack = () => {
    clearSearchedCityReducer();
    inputRef.focus();
  };
  const setInputRef = (searchRef) => {
    inputRef = searchRef;
  };
  return (
    <View style={styles.mainWrapperBox}>
      <SearchBox
        updateSearch={updateSearch}
        searchedKeyword={searchedKeyword}
        isSearchLoading={isSearchLoading}
        setInputRef={setInputRef}
        onFocus={onFocus}
        onClearCallBack={onClearCallBack}
      />

      {isSearchedCityListVisible && <SearchedCityList />}
      {isWeatherDetailsLoading ? (
        <ActivityIndicator
          size={35}
          color={Appearance.getColorScheme() === "dark" ? "white" : "black"}
        />
      ) : (
        areWeatherDetailsVisible && (
          <>
            <CurrentWeatherInfo weatherDetails={weatherDetails} />
            <DropDown
              updateNumberOfDaysForForecast={updateNumberOfDaysForForecast}
              data={FORECAST_DAYS}
              numberOfDaysForForecast={numberOfDaysForForecast}
            />
          </>
        )
      )}

      {isWeatherForecastDetailsLoading ? (
        <ActivityIndicator
          size={35}
          style={{ top: isWeatherDetailsLoading ? 250 : 0 }}
          color={Appearance.getColorScheme() === "dark" ? "white" : "black"}
        />
      ) : (
        areWeatherDetailsVisible && (
          <WeatherForecastDaysList
            weatherForecastDetails={weatherForecastDetails}
          />
        )
      )}
    </View>
  );
};

const mapStateToProps = (state) => ({
  searchedKeyword: state.searchedCityReducer.searchedKeyword,
  searchedCities: state.searchedCityReducer.searchedCities,
  isSearchLoading: state.searchedCityReducer.isSearchLoading,
  selectedCity: state.weatherDetailsOfCityReducer.selectedCity,
  weatherDetails: state.weatherDetailsOfCityReducer.weatherDetails,
  isWeatherDetailsLoading:
    state.weatherDetailsOfCityReducer.isWeatherDetailsLoading,
  weatherForecastDetails:
    state.weatherDetailsOfCityReducer.weatherForecastDetails,
  isWeatherForecastDetailsLoading:
    state.weatherDetailsOfCityReducer.isWeatherForecastDetailsLoading,
  numberOfDaysForForecast:
    state.weatherDetailsOfCityReducer.numberOfDaysForForecast,
});

export default connect(mapStateToProps, {
  fetchSearchedCities,
  updateSearchedKeyword,
  clearSearchedCityReducer,
  updateNumberOfDaysForForecast,
  getWeatherForecastDetailsOfCity,
})(Home);

const styles = StyleSheet.create({
  mainWrapperBox: {
    backgroundColor: defaultBackgroundColor,
    flex: 1,
  },
});
