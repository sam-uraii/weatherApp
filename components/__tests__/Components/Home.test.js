import React from "react";
import renderer from "react-test-renderer";
import configureStore, { Provider } from "redux-mock-store";
import Home from "../../../app/Home";
import SearchBox from "../../SearchBox";
// import { fetchSearchedCities } from "../../../Redux/Middleware/searchedCityMiddleware";
const mockStore = configureStore([]);
jest.mock("../../SearchBox", () => {
  return {
    __esModule: true,
    default: jest.fn(),
  };
});
// jest.mock("../../SearchBox", () => {
//     return React.forwardRef((props, ref) => {
//       return (
//         <TextInput
//           ref={ref} // Attach the ref to the TextInput
//           placeholder="Search..."
//           onChangeText={props.onChangeText}
//         />
//       );
//     });
//   });
SearchBox;
describe("Home Component", () => {
  let store;
  const mockFetchSearchedCities = jest.fn();
  const mockUpdateSearchedKeyword = jest.fn();
  const setInputRef = jest.fn();
  const searchedKeyword = "test";
  const searchedCities = [];
  const selectedCity = null;
  const weatherDetails = {};
  const isWeatherDetailsLoading = false;
  const weatherForecastDetails = [];
  const isWeatherForecastDetailsLoading = false;
  const numberOfDaysForForecast = 1;
  beforeEach(() => {
    store = mockStore({
      searchedCityReducer: {
        searchedKeyword: "",
        searchedCities: [],
        isSearchLoading: false,
      },
      weatherDetailsOfCityReducer: {
        selectedCity: null,
        weatherDetails: {},
        isWeatherDetailsLoading: false,
        weatherForecastDetails: [],
        isWeatherForecastDetailsLoading: false,
        numberOfDaysForForecast: 1,
      },
    });
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        // <Provider store={store}>
        <Home
          fetchSearchedCities={mockFetchSearchedCities}
          updateSearchedKeyword={mockUpdateSearchedKeyword}
          searchedKeyword={searchedKeyword}
          searchedCities={searchedCities}
          selectedCity={selectedCity}
          weatherDetails={weatherDetails}
          isWeatherDetailsLoading={isWeatherDetailsLoading}
          weatherForecastDetails={weatherForecastDetails}
          isWeatherForecastDetailsLoading={isWeatherForecastDetailsLoading}
          numberOfDaysForForecast={numberOfDaysForForecast}
        />
        // </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
