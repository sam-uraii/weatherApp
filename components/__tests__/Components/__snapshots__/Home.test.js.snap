// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#424F6F",
      "flex": 1,
    }
  }
/>
`;
